[SNAP-1427] Use serialization into direct ByteBuffers at Oplog and transport layers

sumwale opened the pull request:
## Changes proposed in this pull request

This set of changes primarily targets two areas creating multiple copies with lots of garbage:

a) Oplog layer: Serialization of incoming objects is done into a flat byte[] then that is copied into a ByteBuffer again which is then written to channel in chunks. This creates unnecessary copies as well as lot of garbage for each write.

b) Connection layer: Like above MsgStreamer and MsgDestreamer classes create unnecessary copies and garbage. The chunking code in these can be replaced with leaner direct channel streaming. The MsgDestreamer class also causes totally avoidable thread-context switches.

Apart from the above, this set of changes also address a long standing problem in GemFire messaging -- the lack of scalability due to per-thread connection model with conserve-sockets=false. The last change is in fact *required* to fix hangs seen in bulk column tables inserts with redundancy in Colo environment.

Oplog changes:
- Added DirectByteBufferDataOutput extending from Kryo's ByteBufferOutput
  to provide a DataOutput on top of an expanding direct ByteBuffer
  (kryo's impl already uses DirectByteBuffer by default). Also corresponding
  DirectByteBufferDataInput to provide a DataInput on top of a DirectByteBuffer.
- Use a new UnsafeHolder.reallocateDirectBuffer above that uses Unsafe.reallocateMemory
  that will do inline expansion if possible. Embed a custom Runnable into the Cleaner
  of DirectByteBuffers that can be reset to not free memory in case reallocation
  was done to expand and create a new DirectByteBuffer.
- Changed the APIs in DiskEntry, OplogSet etc to pass through ValueWrapper which
  has been changed to contain a ByteBuffer instead of a byte[]. In lower level
  internal APIs ByteBuffer is passed directly when "isSerializedObject" flag
  is no longer required.
- Changed Oplog/OverflowOplog to use ChannelBufferUnsafeDataOutputStream instead of
  channel and separate writeBuffer; this existing stream already uses a DirectByteBuffer
  to buffer data, as well as provides a pass-through for large DirectByteBuffers (>16K)
  in its write(ByteBuffer) method so the serialized output in DirectByteBufferDataOutput
  can be directly fed to this avoiding any copying for large enough buffers. Likewise reads
  now directly get a DirectByteBufferDataInput from channel and deserialize region value from it.
  This also simplifies lot of custom serialization, flush code in the two Oplog classes.

P2P connection changes:
- Added MsgChannelStreamer and MsgChannelDestreamer classes that extend
  ChannelBufferUnsafeData* classes to provide a direct ByteBuffer based DataInput/DataOutput.
  MsgChannelStreamer now uses a non-blocking channel to write only as much as possible in
  one go and then move on to writing to other servers (when message has to be sent to multiple
  servers) for best performance.
- New property p2p.useNIOStream (true by default) that switches to use these two classes
  directly on the underlying SocketChannel; for conserve-sockets=true case appropriate locking
  is done in the channels till whole message is sent
- New connection pooling using Apache commons-pool2. A new QueryKeyedObjectPool
  implementation adapted from commons-pool2 GenericKeyedObjectPool has been added that also
  allows for iterating all objects by key etc (where key is a DistributedMember).
- Sending side now uses the above pool with conserve-sockets=false (which has been made the
  default) with pool expiration to keep the total number of connections in check. No selector on
  receiving side since with modern linux NPTL threads, idle threads are no scalability problem except
  for the heap (so selector is only useful to reduce total memory consumption while effort/risk is high).
- P2P Handshake no longer has the additional total size, ID and such fields which are no longer
  useful rather some like total size are not possible to fill in without pre-serializing data fully into buffer
  (or fallback to previous chunking which has other issues like hangs, context switches etc).

## Patch testing

Store and SnappyData precheckin. Extensive TPCH performance testing. Long discussions in email
chains on the gains in GC and performance have been detailed.

## ReleaseNotes changes

The change in the connection model to be pool based from per-thread based with
conserve-sockets=false (which is also the default now).

## Other PRs 

See corresponding PR on SnappyData side.

Commit:
[SNAP-1427] Use serialization into direct ByteBuffers at Oplog layer

- added a DirectByteBufferDataOutput extending from Kryo's ByteBufferOutput
  to provide a DataOutput on top of an expanding direct ByteBuffer
  (kryo's impl already uses DirectByteBuffer by default)
- use a new UnsafeHolder.reallocateDirectBuffer above that uses Unsafe.reallocateMemory
  that will do inline expansion if possible; embed a custom Runnable into the Cleaner
  of DirectByteBuffers that can be reset to not free memory in case reallocation
  was done to expand and create a new DirectByteBuffer
- changed the APIs in DiskEntry, OplogSet etc to pass through ValueWrapper which
  has been changed to contain a ByteBuffer instead of a byte[]; in lower level
  internal APIs ByteBuffer is passed directly when "isSerializedObject" flag
  is no longer required
- changed Oplog to use ChannelBufferUnsafeDataOutputStream instead of channel
  and separate writeBuffer; this existing stream already uses a DirectByteBuffer
  to buffer data, as well as provides a pass-through for large DirectByteBuffers (>16K)
  in its write(ByteBuffer) method so the serialized output in DirectByteBufferDataOutput
  can be directly fed to this avoiding any copying for large enough buffers
- changed OverflowOplog similarly; this also simplifies lot of custom serialization,
  flush code in the two Oplog classes
- release direct ByteBuffers more proactively in SQLBinary (as well as above Oplog classes)
  and not depend on GC to queue up the Cleaner references to minimize overall memory
  pressure on the OS
- removed duplicate release calls in SSLSocketChannel (got pasted twice somehow though does not hurt)

Commit:
create channel during recovery

Commit:
fixing the position of new reallocated buffer in DirectByteBufferDataOutput

Commit:
add back FileChannel variable to Oplogs

previous changes replace FileChannel with Channel*OutputStream which contains
the FileChannel already; however for the read in recovery case no OutputStream
needs to be created, only the FileChannel so using separate channel variable
for FileChannel like before

Commit:
Use direct byte buffers for reading from Oplogs

This is to avoid temporary byte[] objects avoiding both an extra copy as well as possible GC issues.

Changed DirectByteBufferDataOutput.writeUTF to use separate methods from ChannelBufferUnsafeDataOutputStream
since the Kryo writeString is not a compatible format with DataOutput's UTF8 (despite the impl of KryoDataOutput
    which works only for kryo serialization and not elsewhere)

Commit:
missed a conversion point from ByteBuffer to byte[] in GII

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Commit:
align direct buffer allocations to 8 byte word size

Commit:
Avoid logging index size estimation in every run

Skip index size estimation from SnappyRegionStatsCollectorFunction since it is quite expensive.

Commit:
change buffer limit to be incoming size while capacity is aligned one

Commit:
cleaning up OplogFile and re-use in OverflowOplog

Commit:
further cleanup and a bug-fix in compactor for new oplog changes

the issue was length being different from the ByteBuffer.limit() since ByteBuffer.wrap()
should have been with the provided length of byte[] rather than assuming full length

Commit:
correcting the limit/position for OverflowOplog reads

- fixes the BufferOverflow exceptions when reading from unflushed buffer in OVERFLOW cases
- add some of the provided-only dependencies also to testRuntime for GemXD precheckin tests
  that do not refer to SnappyData product jars
- made the fileOffset determination more efficient avoiding Channel.position() call and instead
  using the orginal position + number of bytes written which is already tracked

Commit:
New NIOStream implementation for connection send and receive

- New (default) implementation uses direct streaming over the sockets (i.e. buffered writes)
  rather than chunking to minimize copying and object creation.
- Long standing fix to scalability problem to use pooled connections for sends rather than
  thread-local connections in the new implementation.
  Pooling uses common-pool2 with variant of its default keyed pool, QueryKeyedObjectPool,
  which allows querying, processing objects for each key (aka member in this use case).
- Send connections are non-blocking for best performance where sender can switch sending
  to next server if current one would block. Also corrects hangs seen in blocking mode
  with large number of server-server socket sends of large data on Colo machines.
- Also enabled socket-level native keep-alive options on socket that will detect socket
  failures after some timeout.

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Commit:
Don't skip locks for pooled connections

Skipping the locks is causing some trouble likely because the "sharedResource" flag is
misleading in some code paths. Right now acquiring connection locks in all cases
which will have quite a small overhead when there is no contention
(just one compare-and-swap instruction per connection).

Commit:
Turn conserve-sockets=false by default

With the default connection pooling, the scalability problem with messaging is largely
solved hence turning off conserve-sockets. Having selectors for receivers does not help
in terms of performance in modern Linux distributions using NPTL where there is no overhead
in having idle threads. The only advantage selector brings is lesser amount of memory with
lesser number of threads (which will only matter beyond thousands of nodes). But it also
brings a significant overhead of thread-context switching which can be alleviated by
having a hybrid server like the one implemented for Thrift.

Commit:
Fixing a ClassCastException seen in QueryKeyedObjectPool.foreachObject

Also take the key read lock in foreachObject so that no new key (i.e. member) gets added during iteration

Commit:
allow for incoming buffer in Oplog to have non-zero position

keep writing the buffer to channel until all of it is written (though ideally
    it should be done on first try with FileChannels but no guarantee)

Commit:
fixing an issue of ByteBuffer being read again in Oplog after being read once (for write to channel)

Commit:
increasing max heap for junit tests to 1g due to occasional OOME

Commit:
adding SnappyData modification header to Connection

Commit:
changed QueryKeyedObjectPool to inherit from GenericKeyedObjectPool

now using reflection to get the required fields of GenericKeyedObjectPool
instead of duplicating its code

Commit:
Explicitly mark Connection as pooled

To avoid possible ambiguities with "sharedResource" and "unordered", now explicitly marking
a connection as pooled which is checked when trying to return to the pool.

Use blocking channel mode for non-pooled connections.

Skip the minimum pool size since background thread trying to create connections
(to meet minimum criteria) is causing trouble occasionally especially in startup/shutdown.

Commit:
Moved around ValueWrapper to be inside DiskEntry.Helper

Even though it makes the name a bit verbose (DiskEntry.Helper.ValueWrapper)
it reduces the diffs seen against original code so will help in reviewing/checking

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Commit:
added the condition for SNAP-1410 (PR#163) for remaining cases

also clear the parameters in scrollCursor for last batch in FORWARD_ONLY cursor

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Commit:
allow hostname-for-clients as boot property

Commit:
Increasing log file rollover limit to 1GB

cleaning up the typeIds and methods in Gfxd*Serializable to use message IDs as per convention

Commit:
Fixing build dependencies for store regression tests

Commit:
corrected a comment

hbhanawat commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskEntry.java:
```diff
@@ -614,37 +652,39 @@ else if (value == Token.TOMBSTONE) {
-            bytes = EntryEventImpl.serialize(value);
-            if (bytes.length == 0) {
-              throw new IllegalStateException("serializing <" + value + "> produced empty byte array");
+            buffer = EntryEventImpl.serializeDirect(value, null);
```
I may be missing something but isn't this value going to be DirectByteBuffer for column data? If yes, this could be optimized.

sumwale replied:
Yes, for off-heap storage in SNAP-1454 the value is DirectByteBuffer. That is optimized by:

a) keeping full serialized format of the off-heap value in DirectByteBuffer with data i.e. with the serialization header of typeId, number of bytes

b) BlobHelper method which will check if value is a "SerializedBufferData" then will directly send its buffer as the result

Those changes are already on branch SNAP-1454 (https://github.com/SnappyDataInc/snappy-store/blob/SNAP-1454/gemfire-core/src/main/java/com/gemstone/gemfire/internal/util/BlobHelper.java#L110) and PR for that will be up by tomorrow.

hbhanawat commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/Oplog.java:
```diff
@@ -7865,8 +7995,8 @@ public void initialize(byte opCode,
         Object key = entry.getKeyCopy();
-        this.keyBytes = EntryEventImpl.serialize(key);
-        this.size += (4 + this.keyBytes.length);
+        this.keyBytes = EntryEventImpl.serializeDirect(key, null);
```
Does it help to keep the key on DirectByteBuffer? I remember discussint that small objects can cause fragmentations.

sumwale replied:
Fragmentation is when there are large variations in object sizes. I don't think keys can cause it.

sumwale replied:
Changing keyBytes to use byte[].

hbhanawat commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/DirectByteBufferDataOutput.java:
```diff
@@ -0,0 +1,134 @@
+    final int strLen = s.length();
+    final int utfLen = ClientSharedUtils.getUTFLength(s, strLen);
+    if (utfLen > 65535) {
+      throw new UTFDataFormatException("encoded string too long: " + utfLen
```
Why this size limit when we are letting the buffer to grow until Integer.MAX_VALUE?

sumwale replied:
This is a limitation of DataOutput.writeUTF format that writes the length as a short: https://docs.oracle.com/javase/7/docs/api/java/io/DataOutput.html#writeUTF(java.lang.String). GemFire/GemXD code that intends to write long strings uses DataSerializer.writeString

sumwale replied:
See the code of JDK's DataOutputStream.writeUTF: http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/io/DataOutputStream.java#352

That's primarily where kryo's writeString differs hence this code.

hbhanawat reviewed (approved):
Reviewed the oplog part.

Commit:
incorporating review comments

changed back key serialization in oplog to use byte arrays to reduce possibility of fragmentation in off-heap

add reference to fastutil maps (currently will be used by SnappyData layer but all of store should
    also move from trove maps to fastutil)

Commit:
[SNAP-1488] Fixing couple of issues seen in HA

- Never throw a CacheClosedException from inside fromData. Otherwise partition
  close messages in Cache.close() never get a reply since a previous message on
  the connection had failed corrupting the stream for next message reads
  (and connection is not closed for CacheClosedExceptions due to Gem bug#43543).
  Reviewed all messages to avoid throwing CacheClosedException in fromData rather
  only do it in the processing.

  This issue wasn't seen with older chunking approach because none of these messages
  normally exceeded one chunk (though they could) and that chunk had been entirely
  read before throwing the exception (since it has the size of chunk at start).
- Don't wrap in IllegalStateException if borrowConnection from pool fails due to node
  going down or IOException rather throw the original exception.
- Close connection for any other serialization exceptions in fromData too else
  stream will be corrupt for subsequent reads. This can be a problem with older chunking
  code too if the messages exceed one chunk (but has probably never been tested for that).
- Avoid acquire/release sender semaphore for NIOStream since it is already using a pool connection.

These changes will likely also fix some older (infrequent) hangs reported in HA.

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Commit:
don't ignore CacheClosedException if coming from deserialization

connection should be closed regardless if CacheClosedException is seen in deserialization
since there can be dangling data remaining from the message on the connection

Commit:
correct lock acquire/release for shared connections

Commit:
update sanctionedDataSerializables for changes in ReferencedKeyCheckerFunction, RegionAndKey

Commit:
[SNAP-1503][SNAP-1504] Fix data issue on Oplog rollover

- Caused due to release of the buffer (clearOpState) in an incorrect scope
  where Oplog rollover could lead to recursive call of the method on the new
  Oplog with the same buffer (in Oplog.basicModify).
- Pass around ByteBuffers inside ValueWrapper and BytesAndBits so that on release
  they can be nulled (and code will throw NPE in case of any incorrect buffer releases in future)
- Uniformaly clear the buffer members in DataOutput implementation on a close/release
  so any further incorrect uses of it will throw proper NPEs
- Added DiskStoreImpl.TRACE_READS like the existing TRACE_WRITES to log all reads from
  Oplogs for debugging.
- Removed dependency of DirectByteBufferDataOutput on kryo ByteBufferOutput class
  and use direct buffer methods for simplicity.
- Cleaned up and refactored UTF8 encoding methods in DataOutput implementations
  (ChannelBufferUnsafeDataOutputStream, DirectByteBufferDataOutput, ByteArrayDataOutput)
  to use common methods as far as possible.

Commit:
Reverting some of OverflowOplog to older code ...

... due to some accouting which is not in new code. Don't need to optimize
the overflow-only case so move to older code path. Fixes failures in eviction bts
- added explicit flags to ChannelBufferUnsafe* class constructors to tell whether
  to use Unsafe allocation or allocateDirect; P2P reader sockets, thrift connection
  sockets and value reads (which can be large) use Unsafe while rest use allocateDirect

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Conflicts:
	gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/GfxdSerializable.java

Commit:
[SNAP-1508][SNAP-1509] Hangs in some HA tests

- reason is pool running out of connections; for some reason P2P DirectChannel.sendToMany
  does not mark "inUse=false" for connections where send was unsuccessful for some reason;
  now release the connections to pool no matter whether send was successful or not
- use MAX_PR_THREADS property to also increase the per-server connection pool size correspondingly

Commit:
[SNAP-1488] handle cases of cache going down cleanly

missed a few messages taking handle of GemFireCache and throwing
CacheClosedException in fromData (should never happen)

Commit:
[SNAP-1508][SNAP-1509] try hard to always return connections to pool

Exception cases where getConnection() does not return a connection (even after
    borrowing from pool) etc are now handled to always return connections to pool
regardless of their status (e.g. even if closed)

Commit:
[SNAP-1488] correcting deserialization of AbstractDBSynchronizerMessage

always create event object in deserialization (even if cache is going down)
so that child classes successfully deserialize their data else socket will
have dangling data in cache close and will be closed (leading to infinite
    chain of new connection being created then closed)

Commit:
Use conservative ByteBuffer.allocateDirect for all socket/file channel buffers

Commit:
[SNAP-1508][SNAP-1509] retry pool borrowObject with cancellation check

- Retry QueryKeyedObjectPool borrowObject at fixed intervals till max limit
  is not reached. This helps in two ways:
  - Gets past a quirk in commons-pool2 where it increments created count before
    actual creation, so it can happen many threads incremented the created count
    but all fail in actual creation due to remote node going down. Meanwhile other
    threads start waiting because create count has exceeded max and get stuck.
  - Allows checking for CancelCriterion in retries for the case when node is stopping.
- Moved thread dump method from GfxdLocalLockService to ClientSharedUtils to enable
  usage throughout.

Commit:
update sanctionedDataSerializables.txt for previous change to AbstractDBSynchronizerMessage.fromData

due to change which avoids throwing CacheClosedException in AbstractDBSynchronizerMessage.fromData
the fromData signature has to be updated in sanctionedDataSerializables.txt

Commit:
[SNAP-1514] don't skip isSerializedObject=true flag for off-heap entries

The direct return ValueWrapper for off-heap always assumed isSerializedObject=false
which is incorrect when a raw byte[] data is being stored off-heap.

Commit:
conservatively added null check for ByteBuffer.wrap

Commit:
added RegionEntry.isOffHeap() and changed to use it instead of instanceof OffHeapRegionEntry

in micro-benchmarks the virtual call is 4-5X faster than latter

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1427

Commit:
fix a compilation after previous master merge

Commit:
moving fastutil dependency to gemfire-core (from gemfire-shared) to reduce client jar size

Commit:
correcting a debug log-line for OverflowOplog

sumwale commented:
Merging after correcting all new issues in store regression.
